services/operations.go

```diff
@@ -0,0 +1,430 @@
+package services
+
+import (
+	"fmt"
+
+	"github.com/coinbase/rosetta-sdk-go/types"
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
+	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction/results"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+)
+
+const (
+	// OpTransfer is the Transfer operation.
+	OpTransfer = "Transfer"
+	// OpBurn is the Burn operation.
+	OpBurn = "Burn"
+	// OpReclaimEscrow is the Burn operation.
+	OpReclaimEscrow = "ReclaimEscrow"
+)
+
+// SupportedOperationTypes is a list of the supported operations.
+var SupportedOperationTypes = []string{
+	OpTransfer,
+	OpBurn,
+	OpReclaimEscrow,
+}
+
+const (
+	// OpStatusOK is the operation status for successful operations.
+	OpStatusOK = "OK"
+	// OpStatusFailed is the operation status for failed operations.
+	OpStatusFailed = "Failed"
+)
+
+type transactionsDecoder struct {
+	txs   []*types.Transaction
+	index map[hash.Hash]*types.Transaction
+}
+
+func (d *transactionsDecoder) DecodeTx(rawTx []byte, result *results.Result) error {
+	var sigTx transaction.SignedTransaction
+	if err := cbor.Unmarshal(rawTx, &sigTx); err != nil {
+		return fmt.Errorf("malformed transaction: %w", err)
+	}
+	var tx transaction.Transaction
+	if err := sigTx.Open(&tx); err != nil {
+		return fmt.Errorf("bad transaction signature: %w", err)
+	}
+
+	txHash := sigTx.Hash()
+	rosettaTx := d.getOrCreateTx(txHash)
+
+	// Decode events emitted by the transaction.
+	d.decodeEvents(rosettaTx, result.Events)
+	// In case this transaction failed, there were no events emitted for the failing parts.
+	if !result.IsSuccess() {
+		txSignerAddress := StringFromAddress(staking.NewAddress(sigTx.Signature.PublicKey))
+		o2t := newOperationToTransactionMapper(rosettaTx.Operations)
+		t2o := newTransactionToOperationMapper(&tx, txSignerAddress, OpStatusFailed, rosettaTx.Operations)
+
+		// If no fee operations were emitted, emit some now.
+		if !o2t.HasFee() {
```

Comment:
There are a few cases.

One is:
* Fee operations were already emitted when processing events as the fee was successfully deducted before processing the transaction.
* The transaction failed so no events were emitted and so we need to generate Failed operations.

The other is:
* Fee operations were not emitted when processing events, for example, because there was not enough balance in the account. Failed fee operations need to be emitted here.
* The transaction also failed in this case (since it got aborted when processing the fee), and so we need to generate Failed operations.

Reply:
thanks. could you add this to the comment on line 71?

Reply:
Done.